File: python/pyfory/meta/typedef_decoder.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""
TypeDef decoder for xlang serialization.

This module implements the decoding of TypeDef objects according to the xlang serialization specification.
"""

from typing import List
from pyfory._util import Buffer
from pyfory.meta.typedef import TypeDef, FieldInfo, FieldType
from pyfory.meta.typedef import (
    FieldInfo,
    TypeDef,
    SMALL_NUM_FIELDS_THRESHOLD,
    REGISTER_BY_NAME_FLAG,
    FIELD_NAME_SIZE_THRESHOLD,
    COMPRESS_META_FLAG,
    HAS_FIELDS_META_FLAG,
    META_SIZE_MASKS,
    NUM_HASH_BITS,
    FIELD_NAME_ENCODINGS,
)
from pyfory.type import TypeId
from pyfory.meta.metastring import MetaStringDecoder, Encoding


# Meta string decoders
NAMESPACE_DECODER = MetaStringDecoder(".", "_")
TYPENAME_DECODER = MetaStringDecoder("$", "_")
FIELD_NAME_DECODER = MetaStringDecoder("$", "_")


def decode_typedef(buffer: Buffer, resolver) -> TypeDef:
    """
    Decode a TypeDef from the buffer.

    Args:
        buffer: The buffer containing the encoded TypeDef.
        resolver: The type resolver.

    Returns:
        The decoded TypeDef.
    """
    # Read global binary header
    header = buffer.read_int64()

    # Extract components from header
    meta_size = header & META_SIZE_MASKS
    has_fields_meta = (header & HAS_FIELDS_META_FLAG) != 0
    is_compressed = (header & COMPRESS_META_FLAG) != 0

    # If meta size is at maximum, read additional size
    if meta_size == META_SIZE_MASKS:
        meta_size += buffer.read_varuint32()

    # Read meta data
    meta_data = buffer.read_bytes(meta_size)

    # Decompress if needed
    if is_compressed:
        meta_data = resolver.get_meta_compressor().decompress(meta_data)

    # Create a new buffer for meta data
    meta_buffer = Buffer(meta_data)

    # Read meta header
    meta_header = meta_buffer.read_uint8()

    # Extract number of fields
    num_fields = meta_header & 0b11111
    if num_fields == SMALL_NUM_FIELDS_THRESHOLD:
        num_fields += meta_buffer.read_varuint32()

    # Check if registered by name
    is_registered_by_name = (meta_header & REGISTER_BY_NAME_FLAG) != 0

    # Read type info
    if is_registered_by_name:
        namespace = read_namespace(meta_buffer)
        typename = read_typename(meta_buffer)
        name = namespace + "." + typename if namespace else typename
        # Look up the type_id from namespace and typename
        type_info = resolver.get_typeinfo_by_name(namespace, typename)
        if type_info:
            type_id = type_info.type_id
        else:
            # Fallback to COMPATIBLE_STRUCT if not found
            type_id = TypeId.COMPATIBLE_STRUCT
    else:
        type_id = meta_buffer.read_varuint32()
        type_info = resolver.get_typeinfo_by_id(type_id)
        name = type_info.cls.__name__

    # Read fields info if present
    field_infos = []
    if has_fields_meta:
        field_infos = read_fields_info(meta_buffer, resolver, name, num_fields)

    # Create TypeDef object
    return TypeDef(name, type_id, field_infos, meta_data, is_compressed)


def read_namespace(buffer: Buffer) -> str:
    """Read namespace from the buffer."""
    return read_meta_string(buffer, NAMESPACE_DECODER)


def read_typename(buffer: Buffer) -> str:
    """Read typename from the buffer."""
    return read_meta_string(buffer, TYPENAME_DECODER)


def read_meta_string(buffer: Buffer, decoder: MetaStringDecoder) -> str:
    """Read a meta string from the buffer."""
    # Read encoding and length combined in first byte
    header = buffer.read_uint8()

    # Extract encoding (2 bits) and size (6 bits)
    encoding_value = header & 0b11
    size_value = (header >> 2) & 0b111111

    encoding = Encoding(encoding_value)

    # Read length - same logic as encoder
    length = 0
    if size_value >= FIELD_NAME_SIZE_THRESHOLD:
        length = size_value - FIELD_NAME_SIZE_THRESHOLD + buffer.read_varuint32()
    else:
        length = size_value

    # Read encoded data
    if length > 0:
        encoded_data = buffer.read_bytes(length)
        return decoder.decode(encoded_data, encoding)
    else:
        return ""


def read_fields_info(buffer: Buffer, resolver, defined_class: str, num_fields: int) -> List[FieldInfo]:
    """Read field information from the buffer."""
    field_infos = []
    for _ in range(num_fields):
        field_info = read_field_info(buffer, resolver, defined_class)
        field_infos.append(field_info)
    return field_infos


def read_field_info(buffer: Buffer, resolver, defined_class: str) -> FieldInfo:
    """Read a single field info from the buffer."""
    # Read field header
    header = buffer.read_uint8()

    # Extract field header components
    field_name_encoding = (header >> 6) & 0b11
    field_name_size = (header >> 2) & 0b1111
    if field_name_size == FIELD_NAME_SIZE_THRESHOLD:
        field_name_size += buffer.read_varuint32()
    field_name_size += 1
    encoding = FIELD_NAME_ENCODINGS[field_name_encoding]
    is_nullable = (header & 0b10) != 0
    is_tracking_ref = header & 0b1

    # Read field type info (without flags since they're in the header)
    xtype_id = buffer.read_varuint32()
    field_type = FieldType.xread_with_type(buffer, resolver, xtype_id, is_nullable, is_tracking_ref)

    # Read field name
    field_name = FIELD_NAME_DECODER.decode(buffer.read_bytes(field_name_size), encoding)
    return FieldInfo(field_name, field_type, defined_class)

File: java/fory-core/src/main/java/org/apache/fory/meta/TypeDefDecoder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.meta;

import static org.apache.fory.meta.Encoders.fieldNameEncodings;
import static org.apache.fory.meta.NativeTypeDefDecoder.decodeTypeDefBuf;
import static org.apache.fory.meta.NativeTypeDefDecoder.readPkgName;
import static org.apache.fory.meta.NativeTypeDefDecoder.readTypeName;
import static org.apache.fory.meta.NativeTypeDefDecoder.validateParsedTypeDefHash;
import static org.apache.fory.meta.TypeDef.COMPRESS_META_FLAG;
import static org.apache.fory.meta.TypeDefEncoder.COMPATIBLE_FLAG;
import static org.apache.fory.meta.TypeDefEncoder.FIELD_NAME_SIZE_THRESHOLD;
import static org.apache.fory.meta.TypeDefEncoder.REGISTER_BY_NAME_FLAG;
import static org.apache.fory.meta.TypeDefEncoder.SMALL_NUM_FIELDS_THRESHOLD;
import static org.apache.fory.meta.TypeDefEncoder.STRUCT_FLAG;

import java.util.ArrayList;
import java.util.List;
import org.apache.fory.collection.Tuple2;
import org.apache.fory.exception.DeserializationException;
import org.apache.fory.logging.Logger;
import org.apache.fory.logging.LoggerFactory;
import org.apache.fory.memory.MemoryBuffer;
import org.apache.fory.meta.FieldTypes.FieldType;
import org.apache.fory.meta.MetaString.Encoding;
import org.apache.fory.resolver.TypeInfo;
import org.apache.fory.resolver.XtypeResolver;
import org.apache.fory.serializer.UnknownClass;
import org.apache.fory.type.Types;
import org.apache.fory.util.StringUtils;
import org.apache.fory.util.Utils;

/**
 * A decoder which decode binary into {@link TypeDef}. See spec documentation:
 * docs/specification/fory_xlang_serialization_spec.md <a
 * href="https://fory.apache.org/docs/specification/fory_xlang_serialization_spec">...</a>
 */
class TypeDefDecoder {
  private static final Logger LOG = LoggerFactory.getLogger(TypeDefDecoder.class);

  public static TypeDef decodeTypeDef(XtypeResolver resolver, MemoryBuffer inputBuffer, long id) {
    if ((id & COMPRESS_META_FLAG) != 0) {
      throw new DeserializationException("Compressed xlang TypeDef is not supported");
    }
    Tuple2<byte[], byte[]> decoded = decodeTypeDefBuf(inputBuffer, resolver, id);
    MemoryBuffer buffer = MemoryBuffer.fromByteArray(decoded.f0);
    int header = buffer.readByte() & 0xff;
    boolean isStruct = (header & STRUCT_FLAG) != 0;
    int numFields = 0;
    ClassSpec classSpec;
    if (isStruct) {
      boolean named = (header & REGISTER_BY_NAME_FLAG) != 0;
      boolean compatible = (header & COMPATIBLE_FLAG) != 0;
      int typeId;
      if (named) {
        typeId = compatible ? Types.NAMED_COMPATIBLE_STRUCT : Types.NAMED_STRUCT;
      } else {
        typeId = compatible ? Types.COMPATIBLE_STRUCT : Types.STRUCT;
      }
      numFields = header & SMALL_NUM_FIELDS_THRESHOLD;
      if (numFields == SMALL_NUM_FIELDS_THRESHOLD) {
        int extraFields = buffer.readVarUInt32Small7();
        if (extraFields < 0 || extraFields > Integer.MAX_VALUE - SMALL_NUM_FIELDS_THRESHOLD) {
          throw new DeserializationException("Invalid TypeDef field count");
        }
        numFields += extraFields;
      }
      if (named) {
        String namespace = readPkgName(buffer);
        String typeName = readTypeName(buffer);
        if (Utils.DEBUG_OUTPUT_ENABLED) {
          LOG.info("Decode class {} using namespace {}", typeName, namespace);
        }
        TypeInfo userTypeInfo = resolver.getUserTypeInfo(namespace, typeName);
        if (userTypeInfo == null) {
          classSpec = new ClassSpec(UnknownClass.UnknownStruct.class, typeId, -1);
        } else {
          validateRegisteredTypeDefKind(userTypeInfo, typeId);
          classSpec = new ClassSpec(userTypeInfo.getType(), typeId, userTypeInfo.getUserTypeId());
        }
      } else {
        int userTypeId = buffer.readVarUInt32();
        TypeInfo userTypeInfo = resolver.getUserTypeInfo(userTypeId);
        if (userTypeInfo == null) {
          classSpec = new ClassSpec(UnknownClass.UnknownStruct.class, typeId, userTypeId);
        } else {
          validateRegisteredTypeDefKind(userTypeInfo, typeId);
          classSpec = new ClassSpec(userTypeInfo.getType(), typeId, userTypeId);
        }
      }
    } else {
      if ((header & 0b0111_0000) != 0) {
        throw new DeserializationException("Invalid TypeDef kind header");
      }
      int typeId = nonStructTypeId(header & 0b1111);
      boolean named = Types.isNamedType(typeId);
      if (named) {
        String namespace = readPkgName(buffer);
        String typeName = readTypeName(buffer);
        TypeInfo userTypeInfo = resolver.getUserTypeInfo(namespace, typeName);
        if (userTypeInfo == null) {
          classSpec = new ClassSpec(UnknownClass.UnknownStruct.class, typeId, -1);
        } else {
          validateRegisteredTypeDefKind(userTypeInfo, typeId);
          classSpec = new ClassSpec(userTypeInfo.getType(), typeId, userTypeInfo.getUserTypeId());
        }
      } else {
        int userTypeId = buffer.readVarUInt32();
        TypeInfo userTypeInfo = resolver.getUserTypeInfo(userTypeId);
        if (userTypeInfo == null) {
          classSpec = new ClassSpec(UnknownClass.UnknownStruct.class, typeId, userTypeId);
        } else {
          validateRegisteredTypeDefKind(userTypeInfo, typeId);
          classSpec = new ClassSpec(userTypeInfo.getType(), typeId, userTypeId);
        }
      }
    }
    List<FieldInfo> classFields =
        readFieldsInfo(buffer, resolver, classSpec.entireClassName, numFields);
    if (!isStruct && !classFields.isEmpty()) {
      throw new DeserializationException("Non-struct TypeDef cannot carry field metadata");
    }
    if (buffer.remaining() != 0) {
      throw new DeserializationException("Invalid TypeDef metadata size");
    }
    validateParsedTypeDefHash(id, decoded.f1);
    TypeDef typeDef = new TypeDef(classSpec, classFields, id, decoded.f1);
    if (Utils.DEBUG_OUTPUT_ENABLED) {
      LOG.info("[Java TypeDef DECODED] " + typeDef);
      // Compute and print diff with local TypeDef
      Class<?> cls = classSpec.type;
      if (cls != null && cls != UnknownClass.UnknownStruct.class) {
        TypeDef localDef = TypeDef.buildTypeDef(resolver, cls);
        String diff = typeDef.computeDiff(localDef);
        if (diff != null) {
          LOG.info("[Java TypeDef DIFF] " + classSpec.entireClassName + ":\n" + diff);
        } else {
          LOG.info("[Java TypeDef DIFF] " + classSpec.entireClassName + ": identical");
        }
      }
    }
    return typeDef;
  }

  private static void validateRegisteredTypeDefKind(TypeInfo userTypeInfo, int typeId) {
    int registeredTypeId = userTypeInfo.getTypeId();
    if (registeredTypeId != typeId && !isStructCompatibilityVariant(registeredTypeId, typeId)) {
      throw new DeserializationException(
          String.format(
              "TypeDef kind %s does not match registered kind %s for %s",
              typeId, registeredTypeId, userTypeInfo.getType()));
    }
  }

  private static boolean isStructCompatibilityVariant(int registeredTypeId, int typeId) {
    boolean registeredIdStruct =
        registeredTypeId == Types.STRUCT || registeredTypeId == Types.COMPATIBLE_STRUCT;
    boolean typeIdStruct = typeId == Types.STRUCT || typeId == Types.COMPATIBLE_STRUCT;
    if (registeredIdStruct || typeIdStruct) {
      return registeredIdStruct && typeIdStruct;
    }
    boolean registeredNamedStruct =
        registeredTypeId == Types.NAMED_STRUCT || registeredTypeId == Types.NAMED_COMPATIBLE_STRUCT;
    boolean typeIdNamedStruct =
        typeId == Types.NAMED_STRUCT || typeId == Types.NAMED_COMPATIBLE_STRUCT;
    return registeredNamedStruct && typeIdNamedStruct;
  }

  static int nonStructTypeId(int kindCode) {
    switch (kindCode) {
      case 0:
        return Types.ENUM;
      case 1:
        return Types.NAMED_ENUM;
      case 2:
        return Types.EXT;
      case 3:
        return Types.NAMED_EXT;
      case 4:
        return Types.TYPED_UNION;
      case 5:
        return Types.NAMED_UNION;
      default:
        throw new DeserializationException("Unsupported TypeDef kind code " + kindCode);
    }
  }

  // | header + type info + field name | ... | header + type info + field name |
  private static List<FieldInfo> readFieldsInfo(
      MemoryBuffer buffer, XtypeResolver resolver, String className, int numFields) {
    List<FieldInfo> fieldInfos = new ArrayList<>();
    for (int i = 0; i < numFields; i++) {
      // header: 2 bits field name encoding + 4 bits size + nullability flag + ref tracking flag
      byte header = buffer.readByte();
      int encodingFlags = (header >>> 6) & 0b11;
      boolean useTagID = encodingFlags == 3;
      int fieldNameSize = (header >>> 2) & 0b1111;
      if (fieldNameSize == FIELD_NAME_SIZE_THRESHOLD) {
        fieldNameSize += buffer.readVarUInt32Small7();
      }
      fieldNameSize += 1;
      boolean nullable = (header & 0b10) != 0;
      boolean trackingRef = (header & 0b1) != 0;
      int typeId = buffer.readUInt8();
      FieldType fieldType =
          FieldTypes.FieldType.readCrossLanguage(buffer, resolver, typeId, nullable, trackingRef);

      // read field name or tag ID
      if (useTagID) {
        // When useTagID is true, fieldNameSize actually contains the tag ID
        short tagId = (short) (fieldNameSize - 1);
        // Use a placeholder field name since tag ID is used for identification
        String fieldName = "$tag" + tagId; // TODO we could use id as String as field name
        fieldInfos.add(new FieldInfo(className, fieldName, fieldType, tagId));
      } else {
        Encoding encoding = fieldNameEncodings[encodingFlags];
        String wireFieldName =
            Encoders.FIELD_NAME_DECODER.decode(buffer.readBytes(fieldNameSize), encoding);
        // Convert snake_case field names back to camelCase for Java field matching
        String fieldName = StringUtils.lowerUnderscoreToLowerCamelCase(wireFieldName);
        fieldInfos.add(new FieldInfo(className, fieldName, fieldType));
      }
    }
    return fieldInfos;
  }
}
